File: frontend/src/engine/capabilities/frequencyDistribution.ts
/**
 * frequencyDistribution – Count occurrences of each unique value in a column
 * and write a frequency table.
 *
 * Output columns: Value, Count, Percent (optional).
 * Sorting by value or frequency, ascending or descending.
 */

import { CapabilityMeta, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "frequencyDistribution",
  description: "Count occurrences of each unique value and write a frequency table",
  mutates: true,
  affectsFormatting: false,
};

async function handler(
  context: Excel.RequestContext,
  params: any,
  options: ExecutionOptions,
): Promise<StepResult> {
  const {
    sourceRange,
    outputRange,
    sortBy = "frequency",
    ascending = false,
    includePercent = true,
  } = params as {
    sourceRange: string;
    outputRange: string;
    sortBy?: "value" | "frequency";
    ascending?: boolean;
    includePercent?: boolean;
  };

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would compute frequency distribution of ${sourceRange} → ${outputRange}`,
    };
  }

  options.onProgress?.("Reading source data...");
  const srcRng = resolveRange(context, sourceRange);
  srcRng.load("values");
  await context.sync();

  const vals = (srcRng.values ?? []) as (string | number | boolean | null)[][];
  if (!vals.length) return { stepId: "", status: "success", message: "No data found." };

  options.onProgress?.("Counting frequencies...");

  // Flatten to a single list of values
  const flat: (string | number | boolean)[] = [];
  for (const row of vals) {
    for (const cell of row) {
      if (cell !== null && cell !== "") flat.push(cell);
    }
  }

  // Count frequency (case-insensitive for strings)
  const counts = new Map<string, { display: string | number | boolean; count: number }>();
  for (const val of flat) {
    const key = typeof val === "string" ? val.toLowerCase() : String(val);
    if (!counts.has(key)) {
      counts.set(key, { display: val, count: 0 });
    }
    counts.get(key)!.count++;
  }

  // Convert to sorted array
  const entries = Array.from(counts.values());

  if (sortBy === "frequency") {
    entries.sort((a, b) => ascending ? a.count - b.count : b.count - a.count);
  } else {
    // Sort by value — numeric-aware
    entries.sort((a, b) => {
      const aStr = String(a.display);
      const bStr = String(b.display);
      const cmp = aStr.localeCompare(bStr, undefined, { numeric: true, sensitivity: "base" });
      return ascending ? cmp : -cmp;
    });
  }

  // Build output rows
  const totalCount = flat.length;
  const header: (string | number)[] = includePercent
    ? ["Value", "Count", "Percent"]
    : ["Value", "Count"];

  const dataRows: (string | number)[][] = entries.map((e) => {
    const row: (string | number)[] = [e.display as string | number, e.count];
    if (includePercent) {
      const pct = Math.round((e.count / totalCount) * 10000) / 100; // two decimals
      row.push(pct);
    }
    return row;
  });

  const output = [header, ...dataRows];

  options.onProgress?.("Writing frequency table...");
  const outRng = resolveRange(context, outputRange);
  outRng.getResizedRange(output.length - 1, output[0].length - 1).values = output as any;
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Found ${entries.length} unique values across ${totalCount} total entries`,
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/frequency_distribution.py
"""frequencyDistribution — count occurrences of each unique value.

Flattens the source range, counts each unique value (case-insensitive for
strings), then writes a ``Value / Count [/ Percent]`` table to the output
range. Supports sorting by value or frequency, ascending or descending.
"""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def _natural_key(s: str) -> list[Any]:
    """Numeric-aware sort key mirroring JS localeCompare({numeric: true})."""
    import re

    parts = re.split(r"(\d+)", s)
    key: list[Any] = []
    for part in parts:
        if part.isdigit():
            key.append((0, int(part)))
        else:
            key.append((1, part.lower()))
    return key


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    source = params.get("sourceRange")
    output = params.get("outputRange")
    sort_by = (params.get("sortBy") or "frequency").lower()
    ascending = bool(params.get("ascending", False))
    include_percent = bool(params.get("includePercent", True))

    if not source or not output:
        return {
            "status": "error",
            "message": "frequencyDistribution requires 'sourceRange' and 'outputRange'.",
        }

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": f"Would compute frequency distribution of {source} → {output}.",
        }

    try:
        src_rng = resolve_range(ctx.workbook_handle, source)
        raw = src_rng.value

        # Normalize to 2D.
        if raw is None:
            vals: list[list[Any]] = []
        elif not isinstance(raw, list):
            vals = [[raw]]
        elif raw and not isinstance(raw[0], list):
            rows_, cols_ = src_rng.shape
            vals = [list(raw)] if rows_ == 1 else [[v] for v in raw]
        else:
            vals = [list(r) for r in raw]

        if not vals:
            return {"status": "success", "message": "No data found.", "outputs": {}}

        # Flatten non-null, non-empty cells.
        flat: list[Any] = []
        for row in vals:
            for cell in row:
                if cell is None or cell == "":
                    continue
                flat.append(cell)

        # Count — case-insensitive key for strings.
        counts: dict[str, dict[str, Any]] = {}
        for val in flat:
            key = val.lower() if isinstance(val, str) else str(val)
            if key not in counts:
                counts[key] = {"display": val, "count": 0}
            counts[key]["count"] += 1

        entries = list(counts.values())

        if sort_by == "frequency":
            entries.sort(key=lambda e: e["count"], reverse=not ascending)
        else:  # sort by value — numeric-aware
            entries.sort(
                key=lambda e: _natural_key(str(e["display"])),
                reverse=not ascending,
            )

        total_count = len(flat)
        header: list[Any] = (
            ["Value", "Count", "Percent"] if include_percent else ["Value", "Count"]
        )
        data_rows: list[list[Any]] = []
        for e in entries:
            row: list[Any] = [e["display"], e["count"]]
            if include_percent:
                pct = round((e["count"] / total_count) * 10000) / 100 if total_count else 0
                row.append(pct)
            data_rows.append(row)

        output_rows = [header] + data_rows
        rows = len(output_rows)
        cols = len(header)

        out_rng = resolve_range(ctx.workbook_handle, output)
        out_rng.resize(rows, cols).value = output_rows
    except Exception as exc:  # noqa: BLE001
        return {
            "status": "error",
            "message": f"frequencyDistribution failed: {exc}",
            "error": str(exc),
        }

    return {
        "status": "success",
        "message": f"Found {len(entries)} unique values across {total_count} total entries.",
        "outputs": {"outputRange": output},
    }


registry.register("frequencyDistribution", handler, mutates=True)
